File: backend/app/schemas/tts.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field

from app.db.models import JobStatus


class GenerateRequest(BaseModel):
    model_id: str = Field(..., description="Model to use for generation")
    text: str = Field(..., min_length=1, max_length=4096)
    voice_id: str | None = Field(None, description="Built-in voice slug or voice profile ID")
    speed: float = Field(1.0, ge=0.5, le=2.0)
    pitch: float = Field(0.0, ge=-12.0, le=12.0)
    language: str = Field("en", description="BCP-47 language code")
    output_format: str = Field("wav", pattern="^(wav|mp3|ogg)$")
    extra: dict = Field(default_factory=dict, description="Model-specific parameters")
    keep_alive: int | None = Field(
        None,
        description=(
            "Seconds to keep model in GPU VRAM after generation. "
            "-1 = keep indefinitely, 0 = unload immediately, "
            "None = use server default (MODEL_IDLE_TIMEOUT env var)."
        ),
    )


class GenerateResponse(BaseModel):
    job_id: uuid.UUID
    status: JobStatus
    message: str = "Job queued"


class JobResponse(BaseModel):
    id: uuid.UUID
    model_id: str
    text: str
    voice_id: str | None
    voice_profile_id: uuid.UUID | None
    parameters: dict | None
    status: JobStatus
    error_message: str | None
    output_path: str | None
    duration_seconds: float | None
    processing_time_ms: int | None
    created_at: datetime
    completed_at: datetime | None
    batch_id: uuid.UUID | None = None
    celery_task_id: str | None = None
    speaker_similarity: float | None = Field(
        None,
        description=(
            "Cosine speaker-similarity score in [-1, 1] between the generated "
            "audio and the voice profile's reference clip. NULL when no voice "
            "profile was used or the eval task hasn't run yet. >=0.5 typically "
            "indicates a same-speaker match."
        ),
    )

    model_config = {"from_attributes": True}


class JobListResponse(BaseModel):
    jobs: list[JobResponse]
    total: int
    page: int
    page_size: int


class BatchGenerateRequest(BaseModel):
    lines: list[str] = Field(..., min_length=1, max_length=100)
    model_id: str
    voice_id: str | None = None
    language: str = "en"
    speed: float = Field(1.0, ge=0.5, le=2.0)
    output_format: str = Field("wav", pattern="^(wav|mp3|ogg)$")
    extra: dict = Field(default_factory=dict)


class BatchGenerateResponse(BaseModel):
    batch_id: uuid.UUID
    job_ids: list[uuid.UUID]
    total: int


class BatchStatusResponse(BaseModel):
    batch_id: uuid.UUID
    total: int
    status_counts: dict[str, int]
    jobs: list[JobResponse]

File: frontend/src/lib/api/tts.ts
import axiosInstance from '$lib/axios';

export type JobStatus = 'pending' | 'running' | 'complete' | 'failed' | 'cancelled';

export interface GenerateRequest {
  model_id: string;
  text: string;
  voice_id?: string | null;
  speed?: number;
  pitch?: number;
  language?: string;
  output_format?: 'wav' | 'mp3' | 'ogg';
  extra?: Record<string, unknown>;
}

export interface GenerateResponse {
  job_id: string;
  status: JobStatus;
  message: string;
}

export interface TTSJob {
  id: string;
  model_id: string;
  text: string;
  voice_id: string | null;
  voice_profile_id: string | null;
  parameters: Record<string, unknown> | null;
  status: JobStatus;
  error_message: string | null;
  output_path: string | null;
  duration_seconds: number | null;
  processing_time_ms: number | null;
  created_at: string;
  completed_at: string | null;
  batch_id?: string;
  celery_task_id?: string;
  speaker_similarity?: number | null;
}

export interface JobListResponse {
  jobs: TTSJob[];
  total: number;
  page: number;
  page_size: number;
}

export interface BatchGenerateRequest {
  lines: string[];
  model_id: string;
  voice_id?: string;
  language?: string;
  speed?: number;
  output_format?: 'wav' | 'mp3' | 'ogg';
  extra?: Record<string, unknown>;
}

export interface BatchGenerateResponse {
  batch_id: string;
  job_ids: string[];
  total: number;
}

export interface BatchStatusResponse {
  batch_id: string;
  total: number;
  status_counts: Record<string, number>;
  jobs: TTSJob[];
}

export async function generateTTS(request: GenerateRequest): Promise<GenerateResponse> {
  const res = await axiosInstance.post<GenerateResponse>('/tts/generate', request);
  return res.data;
}

export async function getJob(jobId: string): Promise<TTSJob> {
  const res = await axiosInstance.get<TTSJob>(`/tts/jobs/${jobId}`);
  return res.data;
}

export function getAudioUrl(jobId: string): string {
  return `/api/tts/jobs/${jobId}/audio`;
}

export async function cancelJob(jobId: string): Promise<void> {
  await axiosInstance.delete(`/tts/jobs/${jobId}`);
}

export async function listJobs(params?: {
  page?: number;
  page_size?: number;
  model_id?: string;
  status?: JobStatus;
  search?: string;
}): Promise<JobListResponse> {
  const res = await axiosInstance.get<JobListResponse>('/tts/jobs', { params });
  return res.data;
}

export async function batchGenerate(req: BatchGenerateRequest): Promise<BatchGenerateResponse> {
  const { data } = await axiosInstance.post<BatchGenerateResponse>('/tts/batch', req);
  return data;
}

export async function getBatchStatus(batchId: string): Promise<BatchStatusResponse> {
  const { data } = await axiosInstance.get<BatchStatusResponse>(`/tts/batches/${batchId}`);
  return data;
}

export function getBatchZipUrl(batchId: string): string {
  return `/api/tts/batches/${batchId}/zip`;
}

/** Poll a job until it reaches a terminal state (complete or failed). */
export async function pollJob(
  jobId: string,
  onUpdate: (job: TTSJob) => void,
  intervalMs = 1000,
  timeoutMs = 1_800_000  // 30 min — Fish Speech can take 8+ min for long text
): Promise<TTSJob> {
  const start = Date.now();
  while (Date.now() - start < timeoutMs) {
    const job = await getJob(jobId);
    onUpdate(job);
    if (job.status === 'complete' || job.status === 'failed' || job.status === 'cancelled') {
      return job;
    }
    await new Promise((r) => setTimeout(r, intervalMs));
  }
  throw new Error(`Job ${jobId} timed out after ${timeoutMs}ms`);
}
